miniapps/dpg/util/weakform.hpp

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_DPGWEAKFORM
+#define MFEM_DPGWEAKFORM
+
+#include "mfem.hpp"
+#include "blockstaticcond.hpp"
+
+namespace mfem
+{
+
+/// @brief Class representing the DPG weak formulation.
+class DPGWeakForm
+{
+
+protected:
+
+   BlockStaticCondensation *static_cond; ///< Owned.
+
+   bool initialized = false;
+
+   Mesh * mesh = nullptr;
+   int height, width;
+   int nblocks;
+   Array<int> dof_offsets;
+   Array<int> tdof_offsets;
+
+   /// Block matrix \f$ M \f$ to be associated with the Block bilinear form. Owned.
+   BlockMatrix *mat = nullptr;
+
+   /// BlockVector to be associated with the Block linear form
+   BlockVector * y = nullptr;
+
+   /** @brief Block Matrix \f$ M_e \f$ used to store the eliminations
+        from the b.c.  Owned.
+       \f$ M + M_e = M_{original} \f$ */
+   BlockMatrix *mat_e = nullptr;
+
+   // Trial FE spaces
+   Array<FiniteElementSpace * > trial_fes;
+
+   // Flags to determine if a FiniteElementSpace is Trace
+   Array<int> IsTraceFes;
+
+   // Test FE Collections (Broken)
+   Array<FiniteElementCollection *> test_fecols;
+   Array<int> test_fecols_vdims;
+
+   /// Set of Trial Integrators to be applied for matrix B
+   Array2D<Array<BilinearFormIntegrator * > * > trial_integs;
+
+   /// Set of Test Space (broken) Integrators to be applied for matrix G
+   Array2D<Array<BilinearFormIntegrator * > * > test_integs;
+
+   /// Set of Liniear Froem Integrators to be applied.
+   Array<Array<LinearFormIntegrator * > * > lfis;
+
+   BlockMatrix * P = nullptr; // Block Prolongation
+   BlockMatrix * R = nullptr; // Block Restriction
+
+   mfem::Operator::DiagonalPolicy diag_policy;
+
+   void Init();
+   void ReleaseInitMemory();
+
+   // Allocate appropriate SparseMatrix and assign it to mat
+   void AllocMat();
+
+   void ConformingAssemble();
+
+   void ComputeOffsets();
+
+   virtual void BuildProlongation();
+
+   bool store_matrices = false;
+
+   // Store the matrix L^-1 B  and Vector L^-1 l
+   // where G = L L^t
+   Array<DenseMatrix * > Bmat;
+   Array<Vector * > fvec;
+   Vector residuals;
+
+
+private:
+
+public:
+
+   /// Creates bilinear form associated with FE spaces @a *fespaces.
```

Comment:
I think this commend belongs on the next constructor

Reply:
done!